### PR TITLE
Train image classifier

### PR DESCRIPTION
**Pull request by josephmaa**
beep

### CODE REVIEW

**Comment by josephmaa** [comment]
we want to try harder envs for the cnn, want to see if it can discern cnn -like features

- more bricks (8 width)?
- change kernel/stride to check when breaking?
- bridges with missing brick? 
- bridges that have column/solid stack of bricks?

**Comment by josephmaa** [comment]
lyric's training accuracy to validate for joseph:

training validation accuracy
gap in the middle:  .625  .59
small gap in the middle: .82 .76
is bridge: .95 .99 
less than k: 1 1 
consecutive bricks flat on top: .74 . 74 (goes down to .68 .7 after epoch 4)
one sided: .68 .68 
height: .85 .99 (1 .99, 1 1)

**Comment by josephmaa** [comment]
|                                | Lyric Train Acc | Joseph Train Acc | Lyric Val Acc | Joseph Val Acc |
|--------------------------------|-----------------|------------------|---------------|----------------|
| Gap in the middle              | .625            | .52              | .59           | .59            |
| Small gap in the middle        | .82             | .90              | .76           | .73            |
| Is bridge                      | .95             | 1                | .99           | .99            |
| Less than k                    | 1.              | 1                | 1.            | .99            |
| Consecutive bricks flat on top | .74             | .71              | .74           | .71            |
| One sided                      | .68             | .82              | .68           | .68            |
| Height                         | .85             | 0.85             | .99           | .96            |

**Comment by josephmaa** [comment]
Test different number of fewest common elements on Thurs

**Comment by josephmaa** [comment]
| Effect of class rebalancing to n fewest elements after 300 epochs on eval accuracy |      |      |      |     |
|------------------------------------------------------------------------------------|------|------|------|-----|
|                                                                                    | 5    | 6    | 7    | 10  |
| Bridge height                                                                      | 0.32 | 0.74 | 1.00 | .97 |
| Brick counts                                                                       | 0.16 | 0.11 | 0.14 | .60 |

image

image

**Comment by ldoshi** [comment]
We found a bug in the dim for the softmax.

Now we have a vanishing gradients problem (at least, the gradients are 0 and the loss is correspondingly not changing)

**Comment by josephmaa** [comment]
Test re-weighting of cross entropy loss for multiclass and binary labels on Thursday

**Comment by ldoshi** [comment]
looks like the last layer has a gradient of 0 even if the loss is a positive value. how/why can this happen?

also: try with binary instead of multiclass
try with different loss function

**Comment by josephmaa** [comment]
image
